src/wasm/wasm-binary.cpp

```diff
@@ -4733,12 +4733,15 @@ void WasmBinaryReader::readExports() {
 
 Expression* WasmBinaryReader::readExpression() {
   assert(builder.empty());
-  while (input[pos] != BinaryConsts::End) {
+  while (more() && input[pos] != BinaryConsts::End) {
     auto inst = readInst();
     if (auto* err = inst.getErr()) {
       throwError(err->msg);
     }
   }
+  if (!more()) {
+    throwError("unexpected end of input");
+  }
```

Comment:
```suggestion
  if (!more()) {
    throwError("unexpected end of input");
  }
  while (input[pos] != BinaryConsts::End) {
    auto inst = readInst();
    if (auto* err = inst.getErr()) {
      throwError(err->msg);
    }
  }
```
Wouldn't this be simpler?

Reply:
I think we need to check `more()` on each loop iteration, otherwise it is not valid to read `input[pos]` (as there may be no more to read)? That is, we may run out in the middle of the loop.

Reply:
Ah sorry, you're right, this function reads multiple instructions..